File: web/src/lib/uvIndex.ts
import type { UVPoint } from "./weatherApi";

// WHO UV index bands, coloured like the iOS card.
export const UV_SCALE_MAX = 11;

export type UVBand = {
  label: string;
  color: string;
  limit: number;
};

export const UV_BANDS: readonly UVBand[] = [
  { label: "Low", color: "#70d65c", limit: 3 },
  { label: "Moderate", color: "#e6e040", limit: 6 },
  { label: "High", color: "#f29445", limit: 8 },
  { label: "Very high", color: "#f25257", limit: 11 },
  { label: "Extreme", color: "#cc4de6", limit: Number.POSITIVE_INFINITY },
];

export function uvBand(value: number): UVBand {
  return (
    UV_BANDS.find((band) => value < band.limit) ?? UV_BANDS[UV_BANDS.length - 1]
  );
}

export function uvLevel(value: number): string {
  return uvBand(value).label;
}

export function uvColor(value: number): string {
  return uvBand(value).color;
}

export type UVPeak = { value: number; time: string };

// Highest UV index among the points and the first hour it is reached.
export function uvPeak(points: readonly UVPoint[]): UVPeak | null {
  let peak: UVPeak | null = null;
  for (const point of points) {
    if (!Number.isFinite(point.uv)) continue;
    if (peak == null || point.uv > peak.value) {
      peak = { value: point.uv, time: point.time };
    }
  }
  return peak;
}

export type UVDayBar = {
  hour: number;
  time: string | null;
  uv: number | null;
  past: boolean;
  current: boolean;
};

type LocalTime = { date: string; hour: number };

function localTime(value: Date, timeZone: string): LocalTime {
  const parts = new Intl.DateTimeFormat("en-CA", {
    timeZone,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    hour: "2-digit",
    hour12: false,
  }).formatToParts(value);
  const get = (type: Intl.DateTimeFormatPartTypes) =>
    parts.find((part) => part.type === type)?.value ?? "";
  return {
    date: `${get("year")}-${get("month")}-${get("day")}`,
    hour: Number(get("hour")) % 24,
  };
}

// Points falling on the given local calendar day, one per hour. `offsetDays`
// selects today (0) or tomorrow (1) relative to `now`.
export function uvDayPoints(
  points: readonly UVPoint[],
  timeZone: string,
  now: Date,
  offsetDays = 0,
): UVPoint[] {
  const target = localTime(
    new Date(now.getTime() + offsetDays * 86_400_000),
    timeZone,
  ).date;
  return points.filter(
    (point) => localTime(new Date(point.time), timeZone).date === target,
  );
}

// The 24 hourly slots of today in the local time zone, for the histogram.
// Hours the forecast does not cover have a null value.
export function uvDayBars(
  points: readonly UVPoint[],
  timeZone: string,
  now: Date,
): UVDayBar[] {
  const current = localTime(now, timeZone);
  const bars: UVDayBar[] = Array.from({ length: 24 }, (_, hour) => ({
    hour,
    time: null,
    uv: null,
    past: hour < current.hour,
    current: hour === current.hour,
  }));
  for (const point of uvDayPoints(points, timeZone, now)) {
    const { hour } = localTime(new Date(point.time), timeZone);
    if (Number.isFinite(point.uv)) {
      bars[hour].time = point.time;
      bars[hour].uv = point.uv;
    }
  }
  return bars;
}

// Bar height as a percentage of the scale.
export function uvBarHeight(value: number): number {
  return (Math.min(Math.max(value, 0), UV_SCALE_MAX) / UV_SCALE_MAX) * 100;
}

File: ios/wby/wby/Models/UVIndex.swift
import Foundation
import SwiftUI

/// WHO UV index bands and the per-day slicing of the hourly UV forecast.
enum UVIndex {
    static let scaleMax = 11.0

    struct Band {
        let label: String
        let color: Color
        let limit: Double
    }

    static let bands: [Band] = [
        Band(label: "Low", color: Color(red: 0.44, green: 0.84, blue: 0.36), limit: 3),
        Band(label: "Moderate", color: Color(red: 0.90, green: 0.88, blue: 0.25), limit: 6),
        Band(label: "High", color: Color(red: 0.95, green: 0.58, blue: 0.27), limit: 8),
        Band(label: "Very high", color: Color(red: 0.95, green: 0.32, blue: 0.34), limit: 11),
        Band(label: "Extreme", color: Color(red: 0.80, green: 0.30, blue: 0.90), limit: .infinity),
    ]

    static func band(for value: Double) -> Band {
        bands.first { value < $0.limit } ?? bands[bands.count - 1]
    }

    static func level(for value: Double) -> String {
        band(for: value).label
    }

    static func color(for value: Double) -> Color {
        band(for: value).color
    }

    /// Bar height as a 0–1 fraction of the scale.
    static func barFraction(_ value: Double) -> Double {
        min(max(value, 0), scaleMax) / scaleMax
    }

    struct Peak {
        let value: Double
        let time: Date
    }

    /// Highest UV index among the points and the first hour it is reached.
    static func peak(_ points: [UVPoint]) -> Peak? {
        var peak: Peak?
        for point in points where point.uv.isFinite {
            if let current = peak, point.uv <= current.value { continue }
            peak = Peak(value: point.uv, time: point.time)
        }
        return peak
    }

    /// Points falling on the local calendar day `offsetDays` after `now`.
    static func dayPoints(_ points: [UVPoint], timeZone: TimeZone, now: Date = Date(), offsetDays: Int = 0) -> [UVPoint] {
        var calendar = Calendar.current
        calendar.timeZone = timeZone
        guard let target = calendar.date(byAdding: .day, value: offsetDays, to: now) else { return [] }
        return points.filter { calendar.isDate($0.time, inSameDayAs: target) }
    }

    struct DayBar: Identifiable {
        let hour: Int
        var time: Date?
        var uv: Double?
        let past: Bool
        let current: Bool

        var id: Int { hour }
    }

    /// The 24 hourly slots of today in the local time zone. Hours the forecast
    /// does not cover have a nil value.
    static func dayBars(_ points: [UVPoint], timeZone: TimeZone, now: Date = Date()) -> [DayBar] {
        var calendar = Calendar.current
        calendar.timeZone = timeZone
        let currentHour = calendar.component(.hour, from: now)
        var bars = (0..<24).map { hour in
            DayBar(hour: hour, past: hour < currentHour, current: hour == currentHour)
        }
        for point in dayPoints(points, timeZone: timeZone, now: now) where point.uv.isFinite {
            let hour = calendar.component(.hour, from: point.time)
            bars[hour].time = point.time
            bars[hour].uv = point.uv
        }
        return bars
    }
}
